<div style="width: 100%; clear: both;">
<div style="float: left; width: 50%;">
<img src="http://www.uoc.edu/portal/_resources/common/imatges/marca_UOC/UOC_Masterbrand.jpg" align="left">
</div>
</div>
<div style="float: right; width: 50%;">
<p style="margin: 0; padding-top: 22px; text-align:right;">M2.876 · Análisis de redes complejas</p>
<p style="margin: 0; text-align:right;">20251 - Máster universitario en Ciencias de datos (<i>Data science</i>)</p>
<p style="margin: 0; text-align:right; padding-button: 100px;">Estudios de Informática, Multimedia y Telecomunicaciones</p>
</div>
</div>
<div style="width: 100%; clear: both;">
<div style="width:100%;">&nbsp;</div>

<div class="alert alert-block alert-info">
    <strong>Nombre y apellidos:</strong> Francisco Jesús Castro Garcia
</div>

<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
    <strong>Recordatorio: Verificación de autoría mediante entrevista de contraste </strong>

Esta actividad está sujeta a la verificación de la autoría, la cual puede incluir entrevistas de contraste, tal como se describe en el plan docente de la asignatura. En caso de que el profesorado lo considere necesario, se podrá convocar al estudiante para realizar una entrevista con el objetivo de comprobar la correspondencia entre el contenido presentado y los conocimientos adquiridos, así como la autoría independiente de la actividad.

Encontraréis más información en el plan docente de la asignatura. <br><br>
</div>

<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
    <strong> Fuentes bibliográficas y uso de herramientas de IA. </strong>
Es necesario que el estudiante indique todas las fuentes que ha utilizado para la realización de los diferentes apartados de la PEC y de que forma han sido utilizadas. Esto incluye cualquier uso de herramientas de IA generativa o similares. De no ser así, se considerará que el estudiante ha cometido plagio, siendo penalizado con un suspenso (D) y la posible evaluación negativa de la asignatura de manera íntegra.
    </div>

<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<strong>Figuras y representaciones gráficas</strong> Las figuras deben siempre contener etiqueta en los ejes y ser debidamente claras e informativas. De no cumplirse estas condiciones, las representaciones, cuando se pidan, se considerarán como totalmente incorrectas. Algunas guías sobre como deben ser las figuras:
    
* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4161295/
* https://www.quanthub.com/designing-charts-axes-and-value-labels/
    
Diseñar y representar de forma correcta información gráfica es una competencia transversal que debe ser adquirida.
</div>

<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<strong> Puntualidad en la entrega de la PEC </strong> Las PECs entregadas más allá de la fecha límite (incluso minutos de retraso), serán evaluadas como entrega  realizada en evaluación extraordinaria. Se recomienda, por lo tanto, realizar la entrega con suficiente antelación para prevenir posibles problemas con la exportación del documento, problemas de conexión, etc.  
</div>

# PEC 5: Hackaton en redes complejas



En esta práctica se pondrá en práctica los conocimientos adquiridos en toda la asignatura.

<u>Consideraciones generales</u>: 

- Esta PEC debe realizarse de forma **estrictamente individual**. Cualquier indicio de copia será penalizado con un suspenso (D) para todas las partes implicadas y la posible evaluación negativa de la asignatura de forma íntegra.
- Es necesario que el estudiante indique **todas las fuentes** que ha utilizado para la realización de la PEC. De no ser así, se considerará que el estudiante ha cometido plagio, siendo penalizado con un suspenso (D) y la posible evaluación negativa de la asignatura de forma íntegra.

## Evaluación y entrega (importante leer atentamente)
* La hackaton está alojada en 
https://www.kaggle.com/t/93b960bec781e5e2e6b442fdda727cb7 Deberéis entrar/registraros con vuestro correo de la UOC y vuesto nombre y apellidos. Las instrucciones y los datos se proporcionan en esa hackaton pero a continuación también se tratarán.
* Además hay que entregar en la plataforma el **notebook** junto con el **html** habitual dentro de la fecha de entrega. En este jupyter notebook deberá aparecer vuestra solución. Al final del jupyter notebook deberéis incluir una captura de pantalla con vuestra posición en el ranking del leaderboard de kaggle.
* La nota será la obtenida en el kaggle (ver más abajo), pero esta nota puede ser modificada si el profesor considera que la respuesta no está justificada con el notebook y html entregados.

## Descripción
En esta PEC se proporciona un grafo con todos sus enlaces y las características de cada nodo. Esta red representa las interacciones entre investigadores de distintas áreas de la ciencia de datos. Cada nodo representa a un investigador, mientras que los enlaces indican colaboraciones en proyectos o publicaciones conjuntas.

Además, se proporciona una matriz de características que describe los atributos de cada investigador. De algunos investigadores se conoce claramente cuál es su principal área de trabajo o "etiqueta". El objetivo es predecir la "etiqueta" de los nodos en los que dicha información está desconocida.

Para ello, puedes hacer uso de todos los conocimientos de la asignatura, incluyendo enfoques y técnicas híbridos entre el aprendizaje de grafos, el machine learning, la estadística sobre grafos, la información topológica, etc.

**Leer atentamente debajo el método de evaluación y la forma de enviar los resultados**

## Datos
Se proporcionan los siguentes ficheros (en kaggle):
* El fichero ``node_edges.csv`` contiene los enlaces de toda la red. Se trata de una red **no dirigida**
* El fichero ``node_features.csv`` contiene los features asociados a cada nodo. Cada fila corresponde a un nodo. Cada fila contiene 0 o 1 indicando si la característica está presente.
* ``node_classification_train.csv`` contiene en la primera columna el nombre del nodo y en la segunda columna su clasificación
* ``node_test.csv`` Contiene una columna con el nombre de los nodos de los que hay que predecir su etiqueta.

El fichero de ``node_classification_train`` (o cualquier otro) se puede subdividir como se considere para hacer particiones de train/eval.


## Evaluación
La evaluación consistirá simplemente en una métrica de accuracy sobre los nodos de test. Es decir, para cada nodo contenido en el fichero **node_test.csv** hay que predecir su etiqueta.

**Además hay que subir el jupyter notebook de forma habitual en la página correspondiente del campus de la UOC**

El fichero hay que enviarlo de la siguiente forma:

## Fichero de envio

    id,label
    2,0
    5,0
    6,1
    etc.


## Notas de la PEC.
Dependiendo de la accuracy (a) obtenida, se obtendrán las siguientes puntuaciones:
* a < 56 -> 1 punto
* 56 <a < 58 -> 2 puntos
* 58 <= a < 60 -> 3 puntos
* 60 <= a < 62 -> 4 puntos
* 62 <= a < 64 -> 5 puntos
* 64 <= a < 66 -> 6 puntos
* 66 <= a < 68 -> 7 puntos
* 68 <= a < 70 -> 8 puntos
* a >70 -> 9 puntos
* Las 3 personas con mejor puntuación -> 10 puntos


<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<strong>Solución </strong>
</div>


## Setup Notebook: Carga de Librerias 

In [1]:
# Librerías básicas
import pandas as pd
import numpy as np

# Grafo
import networkx as nx

# Machine Learning
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

# Visualization
import matplotlib.pyplot as plt


## Paso1: Carga de datos

In [2]:
DATA_PATH = "data/"

# =========================
# Edges
# =========================
edges = pd.read_csv(DATA_PATH + "edges.csv")

# =========================
# Node features (NO header)
# =========================
features = pd.read_csv(
    DATA_PATH + "node_features.csv",
    header=None
)

# Crear columna id a partir del índice
features.insert(0, "id", features.index)
features["id"] = features["id"].astype(int)

# =========================
# Train labels
# =========================
train_labels = pd.read_csv(
    DATA_PATH + "node_classification_train.csv",
    header=None
)
train_labels.columns = ["id", "label"]
train_labels["id"] = train_labels["id"].astype(int)

# =========================
# Test nodes 
# =========================
test_nodes = pd.read_csv(
    DATA_PATH + "node_test.csv",
    header=None,      
    names=["id"]
)
test_nodes["id"] = test_nodes["id"].astype(int)

# =========================
# Comprobaciones
# =========================
print("Edges:", edges.shape)
print("Features:", features.shape)
print("Train labels:", train_labels.shape)
print("Test nodes:", test_nodes.shape)



Edges: (5277, 2)
Features: (2708, 479)
Train labels: (992, 2)
Test nodes: (1716, 1)


## Paso2: Construcción del grafo y métricas topológicas

In [3]:
# Construccion de grafo

G = nx.from_pandas_edgelist(
    edges,
    source=edges.columns[0],
    target=edges.columns[1]
)

print(f"Número de nodos: {G.number_of_nodes()}")
print(f"Número de aristas: {G.number_of_edges()}")



Número de nodos: 2708
Número de aristas: 5277


Se incorporan métricas topológicas adicionales para capturar distintas nociones de centralidad e influencia en la red.

In [4]:
# Calcular metricas topologicas

degree_dict = dict(G.degree())
clustering_dict = nx.clustering(G)
pagerank_dict = nx.pagerank(G)

topological_features = pd.DataFrame({
    "id": list(degree_dict.keys()),
    "degree": list(degree_dict.values()),
    "clustering": list(clustering_dict.values()),
    "pagerank": list(pagerank_dict.values())
})

topological_features["id"] = topological_features["id"].astype(int)

In [5]:
# Betweenness centrality
# # Eigenvector centrality

betweenness_dict = nx.betweenness_centrality(G)
eigenvector_dict = nx.eigenvector_centrality(G, max_iter=1000)

topological_features["betweenness"] = topological_features["id"].map(betweenness_dict)
topological_features["eigenvector"] = topological_features["id"].map(eigenvector_dict)

## Paso3: Construcción dataset entrenamiento

In [6]:
# Unir features originales + topológicas
full_features = features.merge(
    topological_features,
    on="id",
    how="left"
)

print(full_features.shape)
full_features.isna().sum().sum()


(2708, 484)


0

In [7]:
# Unir etiquetas
data_labeled = full_features.merge(
    train_labels,
    on="id",
    how="inner"
)
print("data_labeled shape:", data_labeled.shape)
data_labeled.head()

data_labeled shape: (992, 485)


id    0    1    2    3    4    5    6    7    8  ...  474  475  476  477  \
0   0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1   3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
2   4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3   6  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4  11  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

   degree  clustering  pagerank   betweenness   eigenvector  label  
0       3    0.333333  0.000336  9.766154e-07  9.057555e-05      3  
1       1    0.000000  0.000369  0.000000e+00  7.403002e-33      0  
2       5    0.700000  0.000396  5.511293e-04  4.627333e-03      3  
3       4    0.333333  0.000330  6.987223e-04  2.916307e-04      0  
4       2    1.000000  0.000197  0.000000e+00  6.259343e-05      0  

[5 rows x 485 columns]

In [8]:
# Separar X e y
X = data_labeled.drop(columns=["id", "label"])
y = data_labeled["label"]

In [9]:
# Train / validación
X_train, X_val, y_train, y_val = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


In [10]:
print("full_features shape:", full_features.shape)
print("train_labels shape:", train_labels.shape)
print("data_labeled shape:", data_labeled.shape)


full_features shape: (2708, 484)
train_labels shape: (992, 2)
data_labeled shape: (992, 485)


In [11]:
# Escalado (solo para modelos lineales)
X_train.columns = X_train.columns.astype(str)
X_val.columns = X_val.columns.astype(str)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)


## Paso 4: Entrenamiento y evaluación de modelos

### Logistic Regression

In [12]:
# Logistic Regression
log_reg = LogisticRegression(max_iter=1000, random_state=42)
log_reg.fit(X_train_scaled, y_train)

y_val_pred_lr = log_reg.predict(X_val_scaled)
acc_lr = accuracy_score(y_val, y_val_pred_lr)

print("Logistic Regression accuracy:", acc_lr)


Logistic Regression accuracy: 0.5728643216080402


### Random Forest

In [13]:
results = []

for max_depth in [15, 20, 30]:
    for min_leaf in [1, 5, 10]:
        rf = RandomForestClassifier(
            n_estimators=300,
            max_depth=max_depth,
            min_samples_leaf=min_leaf,
            class_weight="balanced",
            random_state=42,
            n_jobs=-1
        )
        rf.fit(X_train, y_train)
        acc = accuracy_score(y_val, rf.predict(X_val))

        results.append({
            "max_depth": max_depth,
            "min_samples_leaf": min_leaf,
            "accuracy": acc
        })

results_df = pd.DataFrame(results)
results_df


max_depth  min_samples_leaf  accuracy
0         15                 1  0.678392
1         15                 5  0.628141
2         15                10  0.603015
3         20                 1  0.683417
4         20                 5  0.638191
5         20                10  0.592965
6         30                 1  0.668342
7         30                 5  0.638191
8         30                10  0.592965

In [14]:
# seleccionar mejor configuración
best_row = results_df.loc[results_df["accuracy"].idxmax()]

best_params = {
    "max_depth": int(best_row["max_depth"]),
    "min_samples_leaf": int(best_row["min_samples_leaf"])
}

best_accuracy = best_row["accuracy"]

best_params, best_accuracy


({'max_depth': 20, 'min_samples_leaf': 1}, 0.6834170854271356)

In [15]:
# Random Forest con la mejor configuración encontrada
rf_best = RandomForestClassifier(
    n_estimators=400,
    max_depth=best_params["max_depth"],
    min_samples_leaf=best_params["min_samples_leaf"],
    class_weight="balanced",
    random_state=42,
    n_jobs=-1
)

rf_best.fit(X_train, y_train)

y_val_pred_rf = rf_best.predict(X_val)
acc_rf = accuracy_score(y_val, y_val_pred_rf)

print("Random Forest (mejor configuración) accuracy:", acc_rf)


Random Forest (mejor configuración) accuracy: 0.6884422110552764


In [16]:
FINAL_MODEL_PARAMS = {
    "n_estimators": 400,              # más árboles = más estabilidad
    "max_depth": best_params["max_depth"],
    "min_samples_leaf": best_params["min_samples_leaf"],
    "class_weight": "balanced",
    "random_state": 42,
    "n_jobs": -1
}

In [17]:
print("Mejor configuración encontrada para Random Forest:")
print(f"  max_depth        = {best_params['max_depth']}")
print(f"  min_samples_leaf = {best_params['min_samples_leaf']}")
print(f"  validation accuracy = {best_accuracy:.4f}")


Mejor configuración encontrada para Random Forest:
  max_depth        = 20
  min_samples_leaf = 1
  validation accuracy = 0.6834


Se realiza una búsqueda controlada de hiperparámetros del modelo Random Forest, analizando la interacción entre la profundidad máxima de los árboles (max_depth) y el número mínimo de muestras por hoja (min_samples_leaf).
Se observa que profundidades intermedias (15–20) y un bajo valor de regularización (min_samples_leaf=1) maximizan la precisión, mientras que valores elevados de regularización conducen a underfitting.

### GradientBoostingClassifier

In [18]:


gb = GradientBoostingClassifier(
    n_estimators=200,
    learning_rate=0.05,
    max_depth=3,
    random_state=42
)

gb.fit(X_train, y_train)
acc_gbc=accuracy_score(y_val, gb.predict(X_val))


### Selección Final Modelos

In [19]:
print("Comparación de modelos:")
analysis_results = {
    "Logistic Regression": acc_lr,
    "Random Forest (tuned)": acc_rf,
    "Gradient Boosting": acc_gbc
}

analysis_results


Comparación de modelos:


{'Logistic Regression': 0.5728643216080402,
 'Random Forest (tuned)': 0.6884422110552764,
 'Gradient Boosting': 0.6582914572864321}

In [20]:
pd.DataFrame.from_dict(
    analysis_results, 
    orient="index", 
    columns=["validation_accuracy"]
)


validation_accuracy
Logistic Regression               0.572864
Random Forest (tuned)             0.688442
Gradient Boosting                 0.658291

In [21]:
## Modelo Final

final_model = RandomForestClassifier(**FINAL_MODEL_PARAMS)


## Paso 5: Entrenamiento final y predicción en test

In [22]:
# Reentrenar con TODOS los datos etiquetados
# Dataset completo etiquetado
X_full = data_labeled.drop(columns=["id", "label"])
y_full = data_labeled["label"]

# Fix scikit-learn
X_full.columns = X_full.columns.astype(str)

#  entrenamiento final
final_model.fit(X_full, y_full)




RandomForestClassifier(class_weight='balanced', max_depth=20, n_estimators=400,
                       n_jobs=-1, random_state=42)

In [32]:
# 1. Construir X_test
X_test = test_nodes.merge(full_features, on="id", how="left")
X_test = X_test.drop(columns=["id"])
X_test.columns = X_test.columns.astype(str)

# 2. Predecir SOLO test
test_predictions = final_model.predict(X_test)

# 3. Crear CSV
submission = pd.DataFrame({
    "id": test_nodes["id"],
    "label": test_predictions
})

submission.to_csv("submission.csv", index=False)



In [33]:
print("Test nodes:", test_nodes.shape)
print("X_test:", X_test.shape)


Test nodes: (1716, 1)
X_test: (1716, 483)


## Conclusiones

En esta práctica se ha abordado un problema de clasificación de nodos en grafos combinando información estructural y atributos de nodo. Tras comparar distintos modelos supervisados y realizar una búsqueda de hiperparámetros, se selecciona Random Forest como modelo final por su equilibrio entre rendimiento, estabilidad e interpretabilidad.

El modelo alcanza una accuracy de validación cercana al 0.65–0.68, lo que indica que la estructura del grafo aporta información relevante para la tarea de clasificación.

## Referencias:

Este notebook ha sido asistido mediante ChatGPT para:

Interpretar enunciados del ejercicio.
Crear funciones y visualizaciones.
Pérez Solà, C. (2023) Detección de comunidades. [Recurso de aprendizaje textual]. 1a edición. Barcelona: Fundació Universitat Oberta de Catalunya (FUOC).

Arola Fernández, Ll. (2023) Difusión y propagación de la información. [Recurso de aprendizaje textual]. 1a edición. Barcelona: Fundació Universitat Oberta de Catalunya (FUOC).

Arola Fernández, L. (2023) Interacciones de orden superior: hipergrafos y complejos simpliciales. [Recurso de aprendizaje textual]. 1.a ed. Barcelona: Fundació Universitat Oberta de Catalunya (FUOC).

Palazzi Nieves, M. J. (2023) Redes multicapa. [Recurso de aprendizaje textual]. 1.a ed. Barcelona: Fundació Universitat Oberta de Catalunya (FUOC).

Casas Roma, J. (2023) Predicción de enlaces. [Recurso de aprendizaje textual]. 1.a ed. Barcelona: Fundació Universitat Oberta de Catalunya (FUOC).

Borondo Benito, J. (2023) Graph Neural Networks: Introducción a las redes neuronales basadas en grafos. [Recurso de aprendizaje textual]. 1.a ed. Barcelona: Fundació Universitat Oberta de Catalunya (FUOC).